python/ray/serve/_private/router.py

```diff
@@ -689,17 +691,142 @@ def _get_singleton_asyncio_loop(cls) -> asyncio.AbstractEventLoop:
+            # It's called when the asyncio Task finishes (success, exception, or cancel).
+            if asyncio_future_callback is not None:
+                asyncio_future_callback(asyncio_future, concurrent_future)
+
+        if asyncio_future_callback is not None:
+            # Add the user-provided callback
+            task.add_done_callback(_asyncio_task_done_callback)
+
+        # Schedule the concurrent_future_callback to run soon in the loop's thread.
+        loop.call_soon_threadsafe(concurrent_future_callback)
+        return concurrent_future
+
     def assign_request(
         self,
         request_meta: RequestMetadata,
         *request_args,
         **request_kwargs,
     ) -> concurrent.futures.Future[ReplicaResult]:
-        return asyncio.run_coroutine_threadsafe(
+        """Schedules assign_request call on the internal asyncio loop.
+
+        This method uses `_run_coroutine_threadsafe` to execute the actual request
+        assignment logic (`_asyncio_router.assign_request`) on the dedicated
+        asyncio event loop thread. It returns a `concurrent.futures.Future` that
+        can be awaited or queried from the calling (potentially synchronous) thread.
+
+        It also sets up a callback to handle cancellation propagation: if the
+        returned `concurrent.futures.Future` is cancelled, it attempts to cancel
+        the underlying asyncio task and the associated `ReplicaResult`.
+
+        Returns:
+            A concurrent.futures.Future resolving to the ReplicaResult representing
+            the assigned request.
+        """
+
+        def asyncio_future_callback(
+            asyncio_future: asyncio.Future, concurrent_future: concurrent.futures.Future
+        ):
+            """Callback attached to the asyncio Task running assign_request.
+
+            This runs when the asyncio Task finishes (completes, fails, or is cancelled).
+            Its primary goal is to propagate cancellation initiated via the
+            `concurrent_future` back to the `ReplicaResult` if possible.
+            """
+            # Check if the cancellation originated from the concurrent.futures.Future
+            if concurrent_future.cancelled():
+                try:
+                    # Check if the asyncio task itself was successfully cancelled.
+                    # .result() will raise CancelledError if it was.
+                    asyncio_future.result()
+
+                    # If .result() did *not* raise CancelledError, it means the task
+                    # finished (successfully or with another exception) *before*
+                    # the cancellation request was fully processed by the asyncio loop.
+                    # We should still attempt to cancel the ReplicaResult object
+                    # obtained from the task's result.
+                    if (
+                        not asyncio_future.cancelled()
+                    ):  # Explicit check might be redundant but clear
```

Comment:
Why do we need this check if we now it didn't throw prior?

Reply:
Actually can we rewrite this to avoid throwing CancelledError and instead check if its cancelled with the predicate?

Reply:
good point.